File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
type Color =
  | "black"
  | "brown"
  | "red"
  | "orange"
  | "yellow"
  | "green"
  | "blue"
  | "violet"
  | "grey"
  | "white";

export function decodedResistorValue(colors: [Color, Color, Color]): string {
  const colorCodes: Record<Color, number> = {
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9,
  };

  const [first, second, third] = colors;
  const baseValue = parseInt(`${colorCodes[first]}${colorCodes[second]}`);
  const multiplier = Math.pow(10, colorCodes[third]);
  const ohmsValue = baseValue * multiplier;

  if (ohmsValue >= 1_000_000_000) {
    return `${ohmsValue / 1_000_000_000} gigaohms`;
  } else if (ohmsValue >= 1_000_000) {
    return `${ohmsValue / 1_000_000} megaohms`;
  } else if (ohmsValue >= 1_000) {
    return `${ohmsValue / 1_000} kiloohms`;
  } else {
    return `${ohmsValue} ohms`;
  }
}

File: solutions/go/resistor-color-trio/1/resistor_color_trio.go
package resistorcolortrio

import "fmt"

var colorMap = map[string]int{
	"black":  0,
	"brown":  1,
	"red":    2,
	"orange": 3,
	"yellow": 4,
	"green":  5,
	"blue":   6,
	"violet": 7,
	"grey":   8,
	"white":  9,
}

func Label(colors []string) string {
	// First two digits give base value
	first := colorMap[colors[0]]
	second := colorMap[colors[1]]
	baseValue := first*10 + second

	// Third color specifies the exponent of 10
	exp := colorMap[colors[2]]

	// Simple integer power of 10
	value := baseValue
	for i := 0; i < exp; i++ {
		value *= 10
	}

	// Scale units in steps of 1000
	unit := "ohms"
	display := value

	if display >= 1_000_000_000 {
		display /= 1_000_000_000
		unit = "gigaohms"
	} else if display >= 1_000_000 {
		display /= 1_000_000
		unit = "megaohms"
	} else if display >= 1_000 {
		display /= 1_000
		unit = "kiloohms"
	}

	return fmt.Sprintf("%d %s", display, unit)
}
